### ISSUE TITLE
panic: vector kind mismatch after coerce

### ISSUE THREAD

**Issue by philrz** [brimdata/super#5463]
Repro is with:

* super commit e69d3fd
* Test data from `s3://brim-sampledata/mgbench/bench1.csup`
* The start of the [bench1/q3 query](https://github.com/brimdata/mgbench-private/blob/main/bench1/q3/q3.spq)

```
$ super -version
Version: v1.18.0-139-ge69d3fd6

$ super dev vector query '
where
    machine_name=="babbage"
    and load_fifteen != null
    and load_five != null
    and load_one != null
    and mem_free != null
    and swap_free != null
    and log_time >= 2017-01-01T00:00:00Z
' bench1.csup

panic: vector kind mismatch after coerce

goroutine 1 [running]:
github.com/brimdata/super/runtime/vam/expr.(*Compare).eval(0xc000911500, {0xc000914040?, 0xc00090e020?, 0xc000101008?})
	/Users/phil/work/super/runtime/vam/expr/compare.go:38 +0x496
github.com/brimdata/super/vector.Apply(0x10?, 0x9d6bb78?, {0xc000914040?, 0x0?, 0x10000101008?})
	/Users/phil/work/super/vector/apply.go:17 +0xa3
github.com/brimdata/super/runtime/vam/expr.(*Compare).Eval(0xc000911500, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/compare.go:23 +0xe6
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54240?, 0xc000911500?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc000911530, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:60 +0x65
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54200?, 0xc000911530?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc000911680, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:55 +0x34
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54200?, 0xc000911680?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc0009117d0, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:55 +0x34
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54200?, 0xc0009117d0?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc000911920, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:55 +0x34
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54200?, 0xc000911920?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc000911a70, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:55 +0x34
github.com/brimdata/super/runtime/vam/expr.EvalBool(0xc00017cc80, {0x9d6bb78, 0xc000800060}, {0x9d54200?, 0xc000911a70?})
	/Users/phil/work/super/runtime/vam/expr/logic.go:110 +0x34
github.com/brimdata/super/runtime/vam/expr.(*And).Eval(0xc000911c50, {0x9d6bb78, 0xc000800060})
	/Users/phil/work/super/runtime/vam/expr/logic.go:55 +0x34
github.com/brimdata/super/runtime/vam/op.(*Filter).Pull(0xc000911c80, 0x0)
	/Users/phil/work/super/runtime/vam/op/filter.go:25 +0x6b
github.com/brimdata/super/runtime/vam.(*Materializer).Pull(0x51db1588?, 0xf8?)
	/Users/phil/work/super/runtime/vam/materialize.go:25 +0x27
github.com/brimdata/super/zbuf.(*pullerReader).Read(0xc000916740)
	/Users/phil/work/super/zbuf/batch.go:193 +0x85
github.com/brimdata/super/zio.CopyWithContext({0x9d6b408, 0xada2760}, {0x51db1588, 0xc0008ff158}, {0x9d51ce0, 0xc000916740})
	/Users/phil/work/super/zio/zio.go:154 +0x5a
github.com/brimdata/super/zio.Copy(...)
	/Users/phil/work/super/zio/zio.go:146
github.com/brimdata/super/cmd/super/dev/vector/query.(*Command).Run(0x9d51cc0?, {0xc00003e2e0, 0x2, 0xc00003e2e0?})
	/Users/phil/work/super/cmd/super/dev/vector/query/command.go:80 +0x5f4
github.com/brimdata/super/pkg/charm.path.run({0xc00071df40, 0x4, 0x4}, {0xc00003e2e0, 0x2, 0x0?})
	/Users/phil/work/super/pkg/charm/path.go:11 +0x7f
github.com/brimdata/super/pkg/charm.(*Spec).Exec(0xad68380, {0xc00003e2b0, 0x5, 0x5})
	/Users/phil/work/super/pkg/charm/charm.go:71 +0x105
main.main()
	/Users/phil/work/super/cmd/super/main.go:41 +0x5b
```

**Comment by mattnibs** [comment]
This appears to be fixed when I ran this on cbb40f53676764ab402ab737b334580834fab450

**Comment by philrz** [comment]
Verified in super commit c06e81a.

I've confirmed via binary search that this was fixed by the changes in #5492.

```
$ super -version
Version: v1.18.0-162-gc06e81af

$ super dev vector query '
where
    machine_name=="babbage"
    and load_fifteen != null
    and load_five != null
    and load_one != null
    and mem_free != null
    and swap_free != null
    and log_time >= 2017-01-01T00:00:00Z
' bench1.csup

{log_time:2017-01-01T00:00:00Z,machine_name:"babbage",machine_group:"Workstations",cpu_idle:99.704834(float32),cpu_nice:0.(float32),cpu_system:0.085(float32),cpu_user:0.035(float32),cpu_wio:0.25(float32),disk_free:113.321(float32),disk_total:146.998(float32),part_max_used:57.(float32),load_fifteen:0.13255(float32),load_five:0.0605(float32),load_one:0.00033333333(float32),mem_buffers:564968.(float32),mem_cached:4564736.(float32),mem_free:2170276.(float32),mem_shared:0.(float32),swap_free:8388604.(float32),bytes_in:9475.592(float32),bytes_out:25.1(float32)}(=bench1)
...
{log_time:2017-01-11T14:32:00Z,machine_name:"babbage",machine_group:"Workstations",cpu_idle:99.8(float32),cpu_nice:0.(float32),cpu_system:0.09428571(float32),cpu_user:0.(float32),cpu_wio:0.1(float32),disk_free:113.318(float32),disk_total:146.998(float32),part_max_used:57.(float32),load_fifteen:0.11890476(float32),load_five:0.08895238(float32),load_one:0.006666667(float32),mem_buffers:567788.(float32),mem_cached:4586100.(float32),mem_free:2143668.(float32),mem_shared:0.(float32),swap_free:8388604.(float32),bytes_in:9327.89(float32),bytes_out:50.25(float32)}(=bench1)
...

Thanks @mattnibs!
